### PR TITLE
add hg - mercurial scm - support

### LINKED ISSUE #88: Mercurial support
**Describe the problem/challenge you have**
We are using quite a lot [mercurial SMC](https://www.mercurial-scm.org/) and it would awesome to be able to clone our projects using vendir.

**Describe the solution you'd like**
It should be less or more a copy of git.go.

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

### PR DESCRIPTION
**Pull request by WnP**
fix #88

### CODE REVIEW

**Comment by WnP** [comment]
ci lint fixed :heavy_check_mark:

File: pkg/vendir/fetch/hg/hg.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+}
+
+func NewHg(opts ctlconf.DirectoryContentsHg,
+	infoLog io.Writer, refFetcher ctlfetch.RefFetcher) *Hg {
+
+	return &Hg{opts, infoLog, refFetcher}
+}
+
+// nolint:golint
+type HgInfo struct {
```
Comment (cppforlife) [inline pkg/vendir/fetch/hg/hg.go:34]: is word "SHA" and "Commit" proper terms within mercurial for those concepts?

File: pkg/vendir/fetch/hg/hg.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+	SHA         string
+	Tags        []string
```
Comment (cppforlife) [inline pkg/vendir/fetch/hg/hg.go]: does not look like we are setting tags within Retrieve. should we add that or remove this field?

File: pkg/vendir/fetch/hg/hg.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+}
+
+func (t *Hg) fetch(dstPath string, tempArea ctlfetch.TempArea) error {
+	authOpts, err := t.getAuthOpts()
+	if err != nil {
+		return err
+	}
+
+	authDir, err := tempArea.NewTempDir("hg-auth")
+	fmt.Printf(">>> authDir: %s", authDir)
```
Comment (cppforlife) [inline pkg/vendir/fetch/hg/hg.go]: nit: debug statement left

File: pkg/vendir/fetch/hg/hg.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+	env := os.Environ()
+
+	hgURL := t.opts.URL
+
+	_, _, err = t.run([]string{"init"}, env, dstPath)
+	if err != nil {
+		return err
+	}
+
+	hgrcPath := filepath.Join(dstPath, ".hg", "hgrc")
```
Comment (cppforlife) [inline pkg/vendir/fetch/hg/hg.go]: per https://github.com/vmware-tanzu/carvel-vendir/issues/88#issuecomment-891986018, we can drop that into separate location

File: pkg/vendir/fetch/hg/hg.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+	if err != nil {
+		return err
+	}
+
+	_, _, err = t.run([]string{"checkout", ref}, env, dstPath)
+
+	return err
+}
+
+func (t *Hg) resolveRef(dstPath string) (string, error) {
```
Comment (cppforlife) [inline pkg/vendir/fetch/hg/hg.go]: doesnt look like this is used. in git impl we are grabbing specific tag before checking out.

File: pkg/vendir/fetch/hg/hg.go (hunk 1 above)
Comment (WnP) [reply pkg/vendir/fetch/hg/hg.go:34]: `SHA` makes sense within hg, but `commits` are named `changeSets`, so `CsetTitle` is surely better here

File: pkg/vendir/fetch/hg/hg.go (hunk 3 above)
Comment (WnP) [reply pkg/vendir/fetch/hg/hg.go]: Ooops :man_facepalming:

File: pkg/vendir/fetch/hg/hg.go (hunk 4 above)
Comment (WnP) [reply pkg/vendir/fetch/hg/hg.go]: totally agree

File: pkg/vendir/fetch/hg/hg.go (hunk 5 above)
Comment (WnP) [reply pkg/vendir/fetch/hg/hg.go]: yes, like I said in my previous comment.

File: pkg/vendir/fetch/hg/hg.go (hunk 1 above)
Comment (cppforlife) [reply pkg/vendir/fetch/hg/hg.go:34]: if you dont mind, lets use "longer" version: ChangeSetTitle, or ChangesetTitle. a bit easier to google around for for mercurial newbies.

File: pkg/vendir/fetch/hg/hg.go (hunk 5 above)
Comment (cppforlife) [reply pkg/vendir/fetch/hg/hg.go]: didnt a see a comment, but saw you removed it.

File: pkg/vendir/fetch/hg/hg.go (hunk 1 above)
Comment (WnP) [reply pkg/vendir/fetch/hg/hg.go:34]: @cppforlife I've rename it to `ChangeSetTitle`.

File: pkg/vendir/fetch/hg/hg.go (hunk 2 above)
Comment (WnP) [reply pkg/vendir/fetch/hg/hg.go]: This is outdated and tags are no longer present.

**Pull request merged**